galaxy_ng/app/api/v3/viewsets/collection.py

```diff
@@ -180,3 +186,64 @@ def get(self, request, *args, **kwargs):
         metrics.collection_artifact_download_failures.labels(status=response.status_code).inc()
         raise APIException('Unexpected response from content app. '
                            f'Code: {response.status_code}.')
+
+
+class CollectionVersionMoveViewSet(ViewSet):
+    permission_classes = GALAXY_PERMISSION_CLASSES + [
+        permissions.IsNamespaceOwner
+    ]
+
+    def move_content(self, request, *args, **kwargs):
+        """Remove content from source repo and add to destination repo.
+
+        Creates new RepositoryVersion of source repo without content included.
+        Creates new RepositoryVersion of destination repo with content included.
+        """
+
+        try:
+            collection_version = CollectionVersion.objects.get(
+                namespace=self.kwargs['namespace'],
+                name=self.kwargs['name'],
+                version=self.kwargs['version'],
+            )
+        except ObjectDoesNotExist:
+            raise NotFound('Collection version not found')
+
+        try:
+            src_repo = AnsibleDistribution.objects.get(
+                base_path=self.kwargs['source_path']).repository
+            dest_repo = AnsibleDistribution.objects.get(
+                base_path=self.kwargs['dest_path']).repository
+        except ObjectDoesNotExist:
+            raise NotFound('Repo(s) for moving collection not found')
+
+        src_versions = CollectionVersion.objects.filter(pk__in=src_repo.latest_version().content)
+        if collection_version not in src_versions:
+            raise NotFound('Collection version not found in source repo')
+
+        dest_versions = CollectionVersion.objects.filter(pk__in=dest_repo.latest_version().content)
+        if collection_version in dest_versions:
+            raise APIException('Collection version already found in destination repo')
+
+        add_task = self._add_content(collection_version, dest_repo)
+        remove_task = self._remove_content(collection_version, src_repo)
+
+        return Response(
+            data={
+                'add_task_id': add_task.id,
+                'remove_task_id': remove_task.id,
+            },
+            status='200'
+        )
+
+    @staticmethod
+    def _add_content(collection_version, repo):
+        locks = [repo]
+        task_args = (collection_version.pk, repo.pk)
+        return enqueue_with_reservation(add_content_to_repository, locks, args=task_args)
+
+    @staticmethod
+    def _remove_content(collection_version, repo):
+        locks = [repo]
+        task_args = (collection_version.pk, repo.pk)
+        return enqueue_with_reservation(remove_content_from_repository, locks, args=task_args)
```

Comment:
Do these tasks change the RepositoryVersion 'atomically'?  Will a /move/ bump the RepositoryVersion twice (once for add, once for remove) ?

Reply:
source_repo gets a new RepositoryVersion, and destination_repo gets a new RepositoryVersion, but this shouldn't be happening on same Repository object.

Reply:
Ah, that makes sense. 

Was going to ask about /copy/, but we discussed it in the pulp meeting. 

(Adding some links to copy impls, mostly just for my reference)

pulp_docker
https://github.com/pulp/pulp_docker/blob/master/pulp_docker/app/tasks/recursive_add.py
https://github.com/pulp/pulp_docker/blob/master/pulp_docker/app/viewsets.py#L317
pulp_rpm
https://github.com/pulp/pulp_rpm/blob/master/pulp_rpm/app/viewsets.py#L250
https://github.com/pulp/pulp_rpm/blob/master/pulp_rpm/app/tasks/copy.py